mesh/ncmesh.hpp

```diff
@@ -51,10 +51,12 @@ struct Embedding
 /// Defines the coarse-fine transformations of all fine elements.
 struct CoarseFineTransformations
 {
-   DenseTensor point_matrices;  ///< matrices for IsoparametricTransformation
+   std::map<Geometry::Type, DenseTensor>
```

Comment:
`std::map` seems like an overkill here. Also, the `point_matrices` member was intended simply as a bank of matrices (with no particular ordering). If the matrices are no longer the same size I think the proper solution would be an array of matrices, indexed by `Embedding:matrix` as before.

Reply:
You may be correct but I don't see how to make that work.  The trouble is that there is only one CoarseFineTransformation object in a Mesh.  If there is only one element type and one refinement strategy then you know how large the point_matrices tensor (or array of DenseMatrix objects) needs to be.  However, if the element type changes then the refinement strategy must change so you need different point matrices.  This much is clear.  The trouble is which point matrices do you need from the unordered array of existing point matrices?  Presumably the order of the point matrices in this array will be determined by the order in which the mesh is traversed (matrices for element 0 followed by the matrices for element 1 etc).  We don't want to duplicate these matrices unnecessarily but we don't know which previously added point matrices are compatible with the current element type.

A map may be overkill but it's seemed like the best choice.  Given that it should never have more than a handful of keys the lookup time should be negligible.  I suppose an alternative would be to store offsets into an array of point matrices and use an indirect index lookup.  This might be complicated if Geometry::Type becomes a char (can you use chars for array indices?)

Reply:
I agree it's difficult to create the matrix "dictionary" without a map and I'm actually also using a map inside `NCMesh::GetRefinementTransforms`, but at the end I dump the matrices into `transforms.point_matrices` and get rid of the map. Thus the map is not propagated outside and the users of `CoarseFineTransformations` can only work with an simple data structure (the structure is basically a sort of interface).

Reply:
I'm sure I don't fully understand how the CoarseFindTransformation is used in the NCMesh class.  Perhaps we should try to discuss this somehow.

Reply:
I think its good to keep the point matrices organized by `Geometry::Type` - a separate `DenseTensor` for each type. Instead of an `std::map`, we can use a simple array of `DenseTensor` but I'm fine with `std::map` as well.

Reply:
Perhaps I've lead us down a rabbit hole by using the enumeration in this way rather than treating it as an `int` or `char`.  We may need to use `std::map` because we really shouldn't use an enumerated type to index into an array.  The values could be negative, they could be non-sequential, they could be `char`.  Technically we would need to create a mapping from the enumerated values to something compatible with `std::size_t` but why bother.

Reply:
You can give the value you want to the values of an enumerated type.
```
enum Geometry { Point = 0, Line = 1, ... };
```

Reply:
Oh you mean if you do `enum Geometry : char`?

Reply:
I realized organizing the matrices by geometry is actually useful to the end-user of the structure, which is the method `FiniteElementSpace::GetLocalRefinementMatrices` (and similar). I think the design should be governed by the usage in these methods and we can return to this when implementing support for mixed meshes in these methods.